### PR TITLE
[WIP] Porting Node changes from Postgres

### PR DESCRIPTION
**Pull request by kavinderd**
This commit just ports the Node changes from the initial CTE commits in
Postgres (44d5be0e5308e951c0c5dc522b4bcacf2bcbc476).

This partial commit is based on the following commit by @hsyuan https://github.com/hsyuan/postgres/commit/be9a01d79fc322050ec75ea0d6fcaaf9e271873b#diff-53078aa83405ab7fc79dfea817fead17

NOTE: This PR is not intended to be merged into master but will first be merged into a separate cte_merge branch

### CODE REVIEW

File: src/backend/optimizer/util/walkers.c [hunk 1]
```diff
@@ -1447,3 +1447,304 @@ find_nodes_walker(Node *node, find_nodes_context *context)
...
+			break;
+		case T_WithClause:
+			return walker(((WithClause *) node)->ctes, context);
+		case T_CommonTableExpr:
+			return walker(((CommonTableExpr *) node)->ctequery, context);
+		default:
+			elog(ERROR, "unrecognized node type: %d",
+				 (int) nodeTag(node));
+			break;
+	}
```
Comment (hlinnaka) [inline src/backend/optimizer/util/walkers.c]: This probably needs to have any GPDB-specific expression types handled, too.

File: src/include/nodes/parsenodes.h [hunk 2]
```diff
@@ -684,18 +698,13 @@ typedef struct RangeTblEntry
...
-	 * denoting columns dropped since the rule was made.
+	 * Fields valid for a CTE RTE (else NULL/zero):
 	 */
-	JoinType	jointype;		/* type of join */
-	List	   *joinaliasvars;	/* list of alias-var expansions */
+	char	   *ctename;		/* name of the WITH list item */
+	Index		ctelevelsup;	/* number of query levels up */
+	bool		self_reference;	/* is this a recursive self-reference? */
+	List	   *ctecoltypes;	/* OID list of column type OIDs */
+	List	   *ctecoltypmods;	/* integer list of column typmods */
```
Comment (vraghavan78) [inline src/include/nodes/parsenodes.h]: folks this is duplicate entry. We already have it in line 708 of the original file

File: src/backend/optimizer/util/walkers.c (hunk 1 above)
Comment (hsyuan) [inline src/backend/optimizer/util/walkers.c]: Yes, we need to add GPDB exprs here. Sooner than later

File: src/backend/nodes/readfuncs.c [hunk 3]
```diff
@@ -598,6 +598,27 @@ _readCommonTableExpr(void)
 }
 
 /*
+ * _readCommonTableExpr
+ */
+static CommonTableExpr *
```
Comment (vraghavan78) [inline src/backend/nodes/readfuncs.c]: this is duplicate code. We already have this. See above.

File: src/backend/nodes/equalfuncs.c [hunk 4]
```diff
@@ -2226,13 +2226,18 @@ _equalRangeTblEntry(RangeTblEntry *a, RangeTblEntry *b)
...
+	COMPARE_SCALAR_FIELD(jointype);
+	COMPARE_NODE_FIELD(joinaliasvars);
 	COMPARE_NODE_FIELD(funcexpr);
 	COMPARE_NODE_FIELD(funccoltypes);
 	COMPARE_NODE_FIELD(funccoltypmods);
 	COMPARE_VARLENA_FIELD(funcuserdata, -1);
 	COMPARE_NODE_FIELD(values_lists);
-	COMPARE_SCALAR_FIELD(jointype);
-	COMPARE_NODE_FIELD(joinaliasvars);
+	COMPARE_STRING_FIELD(ctename);
```
Comment (vraghavan78) [inline src/backend/nodes/equalfuncs.c]: we already have this in line 2247 in the new file. Please remove duplicate

File: src/backend/nodes/outfuncs.c [hunk 5]
```diff
@@ -3927,6 +3961,25 @@ _outRangeFunction(StringInfo str, RangeFunction *node)
 
 #ifndef COMPILING_BINARY_FUNCS
 static void
+_outRangeSubselect(StringInfo str, RangeSubselect *node)
```
Comment (hsyuan) [inline src/backend/nodes/outfuncs.c]: We have this in L3909, dup

File: src/backend/nodes/outfuncs.c [hunk 6]
```diff
@@ -3927,6 +3961,25 @@ _outRangeFunction(StringInfo str, RangeFunction *node)
...
+{
+	WRITE_NODE_TYPE("RANGESUBSELECT");
+
+	WRITE_NODE_FIELD(subquery);
+	WRITE_NODE_FIELD(alias);
+}
+
+static void
+_outRangeFunction(StringInfo str, RangeFunction *node)
```
Comment (hsyuan) [inline src/backend/nodes/outfuncs.c]: We already have this.

File: src/backend/nodes/outfuncs.c [hunk 7]
```diff
@@ -5040,6 +5102,12 @@ _outNode(StringInfo str, void *obj)
 			case T_RangeFunction:
 				_outRangeFunction(str, obj);
 				break;
+			case T_RangeSubselect:
+				_outRangeSubselect(str, obj);
+				break;
+			case T_RangeFunction:
+				_outRangeFunction(str, obj);
+				break;
```
Comment (hsyuan) [inline src/backend/nodes/outfuncs.c]: This is duplicate, see above lines

File: src/backend/optimizer/util/walkers.c (hunk 1 above)
Comment (vraghavan78) [inline src/backend/optimizer/util/walkers.c]: Do we have such a list?

**Comment by kavinderd** [comment]
Merged into branch `cte_merge`

**Pull request closed without merging**